src/xenia/base/clock_posix.cc

```diff
@@ -7,40 +7,51 @@
  ******************************************************************************
  */
 
-#include "xenia/base/clock.h"
-
 #include <sys/time.h>
-#include <time.h>
+
+#include "xenia/base/assert.h"
+#include "xenia/base/clock.h"
 
 namespace xe {
 
 uint64_t Clock::host_tick_frequency_platform() {
   timespec res;
-  clock_getres(CLOCK_MONOTONIC_RAW, &res);
+  auto error = clock_getres(CLOCK_MONOTONIC_RAW, &res);
+  assert_true(!error);
+  assert_true(res.tv_sec == 0); // Sub second resolution is required.
 
-  return uint64_t(res.tv_sec) + uint64_t(res.tv_nsec) * 1000000000ull;
+  // Convert nano seconds to hertz. Resolution is 1ns on most systems.
+  return 1000000000ull / res.tv_nsec;
 }
 
 uint64_t Clock::host_tick_count_platform() {
-  timespec res;
-  clock_gettime(CLOCK_MONOTONIC_RAW, &res);
+  timespec tp;
+  auto error = clock_gettime(CLOCK_MONOTONIC_RAW, &tp);
+  assert_true(!error);
 
-  return uint64_t(res.tv_sec) + uint64_t(res.tv_nsec) * 1000000000ull;
+  return tp.tv_nsec + tp.tv_sec * 1000000000ull;
 }
 
 uint64_t Clock::QueryHostSystemTime() {
-  struct timeval tv;
-  gettimeofday(&tv, NULL);
-
-  uint64_t ret = tv.tv_usec;
-  ret /= 1000;  // usec -> msec
-
-  ret += (tv.tv_sec * 1000);  // sec -> msec
-  return ret;
+  // https://docs.microsoft.com/en-us/windows/win32/sysinfo/converting-a-time-t-value-to-a-file-time
+  constexpr uint64_t seconds_per_day = 3600 * 24;
+  // Don't forget the 89 leap days.
+  constexpr uint64_t seconds_1601_to_1970 =
+      ((369 * 365 + 89) * seconds_per_day);
+
+  timeval now;
+  auto error = gettimeofday(&now, nullptr);
+  assert_true(!error);
+
+  // NT systems use 100ns intervals.
+  return static_cast<uint64_t>(
+             (static_cast<int64_t>(now.tv_sec) + seconds_1601_to_1970)) *
+             10000000ull +
+         now.tv_usec * 10;
```

Comment:
tv_usec is also signed, the final result needs to be converted.

Reply:
wtf yes